lib/transaction/priority_queue_test.go

```diff
@@ -270,3 +272,100 @@ func TestRemoveExtrinsic(t *testing.T) {
 		t.Fatalf("Fail: got %v expected %v", res, tests[1])
 	}
 }
+
+func Test_PopWithTimer(t *testing.T) {
+	pq := NewPriorityQueue()
+	slotTimer := time.NewTimer(time.Second)
+
+	tests := []*ValidTransaction{
+		{
+			Extrinsic: []byte("a"),
+			Validity:  &Validity{Priority: 1},
+		},
+		{
+			Extrinsic: []byte("b"),
+			Validity:  &Validity{Priority: 4},
+		},
+		{
+			Extrinsic: []byte("c"),
+			Validity:  &Validity{Priority: 2},
+		},
+		{
+			Extrinsic: []byte("d"),
+			Validity:  &Validity{Priority: 17},
+		},
+		{
+			Extrinsic: []byte("e"),
+			Validity:  &Validity{Priority: 2},
+		},
+	}
+
+	expected := []int{3, 1, 2, 4, 0}
+
+	for _, test := range tests {
+		pq.Push(test)
+	}
+
+	counter := 0
+	for {
+		txn := pq.PopWithTimer(slotTimer)
+		if txn == nil {
+			break
+		}
+		assert.Equal(t, tests[expected[counter]], txn)
+		counter++
+	}
+}
+
+func Test_PopWithTimer_Ends(t *testing.T) {
+	pq := NewPriorityQueue()
+	// increase sleep time greater than timer
+	pq.pollInterval = 2 * time.Second
+	slotTimer := time.NewTimer(time.Second)
```

Comment:
nit it would be nice to have this test take a shorter time and cover more cases, you can do that using the :warning: change mentioned above, and test for:

- empty queue polled once (at first if block, no ticker)
- empty queue polled multiple times (twice or more, you can set the period to 1ns)
- queue with one element polled once (at first if block, no ticker)
- empty queue polled multiple times until new element is added.. period can also be 1ns, but you need to add an element after i.e. 1ms.

I wrote such test for contexts, and adapted it for PopWithTimer, and it should cover all code paths, cases above and take up to 1ms (I didn't try it though!):

```go
func Test_PriorityQueue_PopWithTimer(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		queueBuilder  func() *PriorityQueue
		queueModifier func(queue *PriorityQueue, done chan<- struct{})
		timer         *time.Timer
		transaction   *ValidTransaction
	}{
		"empty queue polled once": {
			// test should last 1ns
			queueBuilder: NewPriorityQueue,
			timer:        time.NewTimer(time.Nanosecond),
		},
		"empty queue polled multiple times": {
			// test should last 1ms
			queueBuilder: func() *PriorityQueue {
				queue := NewPriorityQueue()
				queue.pollInterval = time.Nanosecond
				return queue
			},
			timer: time.NewTimer(time.Millisecond),
		},
		"queue with one element polled once": {
			// test should be instantaneous
			queueBuilder: func() *PriorityQueue {
				queue := NewPriorityQueue()
				queue.Push(&ValidTransaction{Validity: &Validity{Priority: 1}})
				return queue
			},
			timer:       time.NewTimer(time.Nanosecond),
			transaction: &ValidTransaction{Validity: &Validity{Priority: 1}},
		},
		"queue polled multiple times until new element": {
			// test should last 1ms
			queueBuilder: func() *PriorityQueue {
				queue := NewPriorityQueue()
				queue.pollInterval = time.Nanosecond
				return queue
			},
			queueModifier: func(queue *PriorityQueue, done chan<- struct{}) {
				close(done)
				time.Sleep(time.Millisecond)
				queue.Push(&ValidTransaction{Validity: &Validity{Priority: 1}})
			},
			timer:       time.NewTimer(time.Second),
			transaction: &ValidTransaction{Validity: &Validity{Priority: 1}},
		},
	}

	for name, testCase := range testCases {
		testCase := testCase
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			queue := testCase.queueBuilder()

			modifyDone := make(chan struct{})
			if testCase.queueModifier != nil {
				// modify queue asynchronously while popping
				go testCase.queueModifier(queue, modifyDone)
			} else {
				close(modifyDone)
			}

			transaction := queue.PopWithTimer(testCase.timer)
			<-modifyDone
			testCase.timer.Stop()
			assert.Equal(t, testCase.transaction, transaction)
		})
	}
}

```